pkg/util/limiter/query_limiter.go

```diff
@@ -0,0 +1,78 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+type queryLimiterCtxKey struct{}
+
+var (
+	qlCtxKey        = &queryLimiterCtxKey{}
+	errMaxSeriesHit = "The query hit the max number of series limit while fetching chunks (limit: %d)"
+)
+
+type QueryLimiter struct {
+	uniqueSeriesMx sync.RWMutex
```

Comment:
Given that Read-lock is only used from tests and not production code, I'd suggest to use `sync.Mutex`, which is faster than `sync.RWMutex`. (In your benchmark by about 7.5%)